src/uwtools/config/core.py

```diff
@@ -233,69 +278,23 @@ def dereference_all(self) -> None:
             self.dereference()
             prev = copy.deepcopy(self.data)
 
-    def dictionary_depth(self, config_dict: dict) -> int:
-        """
-        The depth of a dictionary.
-
-        :param config_dict: The dictionary whose depth to find.
-        """
-        if isinstance(config_dict, dict):
-            return 1 + (max(map(self.dictionary_depth, config_dict.values())))
-        return 0
-
-    def iterate_values(
-        self,
-        config_dict: dict,
-        set_var: List[str],
-        jinja2_var: List[str],
-        empty_var: List[str],
-        parent: str,
-    ) -> None:
-        """
-        Characterize values as complete, empty, or Jinja2 templates.
-
-        complete -> set_var, empty -> empty_var, templates -> jinja2_var
-
-        :param config_dict: The dictionary to examine.
-        :param set_var: Complete values.
-        :param jinja2_var: Jinja2 template values.
-        :param empty_var: Empty values.
-        :param parent: Parent key.
-        """
-        for key, val in config_dict.items():
-            if isinstance(val, dict):
-                set_var.append(f"    {parent}{key}")
-                new_parent = f"{parent}{key}."
-                self.iterate_values(val, set_var, jinja2_var, empty_var, new_parent)
-            elif isinstance(val, list):
-                set_var.append(f"    {parent}{key}")
-                for item in val:
-                    if isinstance(item, dict):
-                        self.iterate_values(item, set_var, jinja2_var, empty_var, parent)
-            elif "{{" in str(val) or "{%" in str(val):
-                jinja2_var.append(f"    {parent}{key}: {val}")
-            elif val == "" or val is None:
-                empty_var.append(f"    {parent}{key}")
-            else:
-                set_var.append(f"    {parent}{key}")
-
+    @staticmethod
     @abstractmethod
-    def dump_file(self, path: str) -> None:
+    def dump_dict(path: OptionalPath, cfg: dict, opts: Optional[ns] = None) -> None:
         """
-        Dumps the config as a file.
+        Dumps a provided config dictionary as a file.
 
         :param path: Path to dump config to.
+        :param cfg: The in-memory config object to dump.
+        :param opts: Other options required by a subclass.
         """
 
-    @staticmethod
     @abstractmethod
-    def dump_file_from_dict(path: str, cfg: dict, opts: Optional[ns] = None) -> None:
+    def dump_file(self, path: DefinitePath) -> None:
```

Comment:
The names do seem bad. But here are the facts:

`dump_file()` is an instance method that dumps THIS object's config to a file. Its name hasn't changed.

`dump_dict()` dumps to some IO stream -- it might be a file, it might be `stdout`. It's a _static_ method that will dump whatever `dict` is passed in, in the format managed by that class (Fortran namelist if `NMLConfig.dump_dict()` is called, YAML if `YAMLConfig.dump_dict()` is called, etc.) It used to be called `dump_file_from_dict()`, but "file" is no longer accurate here and, honestly, I don't think `dict` belongs in the name -- that's type information.

Thoughts on better names for these? Maybe `dump_file()` is ok if we clarify the other? Maybe... `dump_formatted()` or something, i.e. we're formatting the given `dict` and then dumping that?

Reply:
Would `dump_self` and `dump_dict` given that "dump" already implies that you're sending it to an output stream make sense? 

I'm still mainly concerned that the two function names are symmetrical. One is "dump [self] to file" and the other is "dump dict [to output stream]". In other words, one is named for where it's being dumped with the what implied, and the other is named for what is being dumped with the where implied. If we could get them both into a "what are we dumping" state, it might be clearer to the user.

Reply:
I like `dump_self` and `dump_dict`. I'll update to use those. I do think those are clearer.

Reply:
Oooh. One more wrench. I'm sorry. Would `dump` and `dump_dict` make sense? Just thinking of how the calls to them would read.

```
NMLConfig.dump_dict() 
config_obj = YAMLConfig()
config_obj.dump()
```

Does it make sense for it to read:

```
config_obj.dump_self()
```

Reply:
Yes, that's even better. Will update `dump_self` -> `dump`.